### PR TITLE
check uploaded twitter and facebook image requirements

### LINKED ISSUE #212: Check whether uploaded Twitter image meets requirements
## Check requirements

When the user uploads an image for the Twitter preview, this image must meet the following requirements (see [here](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/summary-card-with-large-image) and [here](https://developer.twitter.com/en/docs/tweets/optimize-with-cards/overview/summary)):

* Images must be less than 5MB in size. 
* JPG, PNG, WEBP and GIF formats are supported. 
* Only the first frame of an animated GIF will be used. 
* SVG is not supported.
* Dimensions for `summary_large_image` card: minimum 300x157, maximum 4096x4096.
* Dimensions for `summary` card: minimum ~~144x144~~ 200x200, maximum 4096x4096.

## Show notification if requirement isn't met
If a file is uploaded to the media library that does not meet these requirements, a notification should appear in the preview (this was specified in the [roadmap](https://github.com/Yoast/roadmap/issues/21)):

54279211-ca1ec000-4594-11e9-8e38-023225e3365c

**Instead of the notification shown above, use the already existing Alert component (type: warning).**

The text of the notification should be adjusted to the particular requirement(s) that wasn't / weren't met, see [below](https://github.com/Yoast/javascript/issues/212#issuecomment-483148902). A notification should also be shown if an image meets the requirements, but fails to load for other reasons. 

If multiple requirements are not met, combine all warnings into one alert, starting with the following sentence "The uploaded image has the following issue(s):"

## Technical decisions

* If multiple requirements aren't met, we show multiple notifications, one for each unmet requirement.
* Currently, `TwitterImage.js` contains its own checks for failing images, but these will be removed in #127.

## To do
- [ ] Implement the requirement checks.
- [ ] Show a warning if a requirement is not met.
- [ ] Write tests for the requirement checks.
- [ ] Update/create a snapshot for the component. Don't forget to test the case where multiple requirements are not met.

### LINKED ISSUE #214: Check whether uploaded Facebook image meets requirements
## Check requirements

When the user uploads an image for the Facebook preview, this image must meet the following requirements:

* The minimum dimensions for the image are 200x200 px.

## Show notification if requirement isn't met
If a file is uploaded to the media library that does not meet this requirement, a notification should appear in the preview (this was specified in the [roadmap](https://github.com/Yoast/roadmap/issues/21)):

54279211-ca1ec000-4594-11e9-8e38-023225e3365c

A notification should also be shown if an image meets the requirements, but fails to load for other reasons. 

Of course, the notification should say "Facebook" rather than Twitter", and the text of the notification should be adjusted to the particular reason for failure (see [here](https://github.com/Yoast/javascript/issues/214#issuecomment-483149837)).

**Instead of the notification shown above, use the already existing Alert component (type: warning).**

## Technical decisions

* Currently, `FacebookImage.js` contains its own checks for failing images, but these will be removed in #127.
~~* In the old implementation, the minimum image dimensions are 200x200 (see `yoast-social-previews/js/facebookPreview.js`), but in the `FacebookImage` component (in `javascript/packages/social-metadata-previews`), the minimum image size is set to 158x158. The specs should be clear before someone starts working on this issue.~~

## To do
- [ ] Implement the requirement checks.
- [ ] Show a warning if a requirement is not met.
- [ ] Write tests for the requirement checks.
- [ ] Update/create a snapshot for the component.

### PR DESCRIPTION
**Pull request by GuidoOffermans**
## Summary

This PR can be summarized in the following changelog entry:

* Add validationFactory. This runs the provided validation checks. 
* Add facebook and twitter validation checks.
* Add tests for all these validation checks.

## Relevant technical choices:

* Created an image object in the initial state. And moved imageType and imageUrl to this object.
   This makes sure all the information about the image is available in one place.
* Resolved issue #214 in this PR because there is overlapping logic.
* Created 'twitterValidation', 'facebookValidation' and 'validationFactory' in `packages/helpers` because they should be available in the free and premium plugin.

## Test instructions

This PR can be tested by following these steps:

* There a test written, check if these resolve correctly.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  * 

## UI changes
* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Quality assurance

* [ ] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #212 
Fixes #214

### CODE REVIEW

File: packages/helpers/src/social-preview-image-validation/twitterValidation.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ * @returns {String | Boolean} A warning string | true.
+ */
+const validateSize = ( image ) => {
+	const { width, height } = image;
+
+	const warningMessage = sprintf(
+		/* Translators: %d expands to the minimum width, %d expands to the minimum hight,
+		%d expands to the maximum width, %d expands to the maximum hight. */
+		__(
+			"Your image dimensions are not suitable: The minimum dimensions are %dx%d pixels. The maximum dimensions are %dx%d pixels.",
```
Comment (bintzandt) [inline packages/helpers/src/social-preview-image-validation/twitterValidation.js]: Did you check this copy with Thijs/Jessie? It differs from the copy provided in the issue.

If they are okay with this, I would suggest changing the `:` to a `.` so we start a new sentence.

File: packages/helpers/tests/social-preview-image-validation/facebookValidationTest.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		const actual = validateFacebookImage( testImage );
+
+		const expected = [
+			"Your image dimensions are not suitable: The minimum dimensions are 200x200 pixels." ];
+
+		expect( actual ).toEqual( expected );
+	} );
+
+	// Image type
+	it(	"returns an array with a warning about: The uplaoded image is not supported; when the fileType is not jpeg, gif or png", () => {
```
Comment (bintzandt) [inline packages/helpers/tests/social-preview-image-validation/facebookValidationTest.js]:
```suggestion
	it(	"returns an array with a warning about: The uploaded image is not supported; when the fileType is not jpeg, gif or png", () => {
```

File: packages/helpers/src/social-preview-image-validation/validationFactory.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+			if ( result === true ) {
+				return;
+			}
+
+			warnings.push( result );
+		} );
+		return warnings;
+	};
+};
```
[the remark this answers is not in the document]
Comment (GuidoOffermans) [reply packages/helpers/src/social-preview-image-validation/validationFactory.js]: I also like the first option. Clean and simple.

**Comment by Xyfi** [comment]
CR 👍 
Acceptance 👍

**Pull request merged**